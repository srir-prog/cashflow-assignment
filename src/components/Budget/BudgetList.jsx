import React from "react";
import IncomeList from "./IncomeList";
import ExpenseList from "./ExpenseList";
import { useTranslation } from "react-i18next";

const BudgetList = ({ allItems }) => {
  //For internationalisation.
  const [t] = useTranslation("common");
  const { income: incomes, expense: expenses } = allItems;
  return (
    <div className="container clearfix">
      <div className="income">
        <h2 className="icome__title">{t("bottom.income")}</h2>
        <div className="income__list">
          {incomes.length > 0 &&
            incomes.map((income) => (
              <IncomeList key={income.id} income={income} />
            ))}
        </div>
      </div>
      <div className="expenses">
        <h2 className="expenses__title">{t("bottom.expenses")}</h2>
        <div className="expenses__list">
          {expenses.length > 0 &&
            expenses.map((expense) => (
              <ExpenseList key={expense.id} expense={expense} />
            ))}
        </div>
      </div>
    </div>
  );
};

export default BudgetList;
