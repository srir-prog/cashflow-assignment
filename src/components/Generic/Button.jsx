import React from "react";
const Button = ({ onClick, buttonIcon, className }) => {
  return (
    <button className={className} onClick={onClick}>
      <i className={buttonIcon}></i>
    </button>
  );
};

export default Button;
