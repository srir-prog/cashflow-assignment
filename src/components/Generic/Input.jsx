import React from "react";
const Input = ({ value, onChange, type, placeholder, className, name }) => {
  return (
    <input
      name={name}
      type={type}
      className={className}
      placeholder={placeholder}
      value={value}
      onChange={onChange}
    />
  );
};

export default Input;
