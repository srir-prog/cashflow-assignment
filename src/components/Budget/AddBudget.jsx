import React from "react";
import Select from "../Generic/Select";
import Button from "../Generic/Button";
import Input from "../Generic/Input";
import {
  addItemToStore,
  addTotalIncome,
  addDataForSankeyChart,
} from "../../store/budgetSlice";
import { useDispatch, useSelector } from "react-redux";
import BudgetList from "./BudgetList";
import { useTranslation } from "react-i18next";

const AddBudget = () => {
  //For internationalisation.
  const [t] = useTranslation("common");
  //Form a state object to store the values locally
  const [formData, setFormData] = React.useState({
    description: "",
    cost: "",
    type: "income",
  });

  //Destructuring form data.
  const { type, description, cost } = formData;

  //Making use of useDispatch hook from
  //"react-redux" to expose a dispatch function.
  const dispatch = useDispatch();

  //Making use of useSelector hook from
  //"react-redux" to access reuired state from the store
  const { allItems, totalIncome } = useSelector((state) => ({
    allItems: state.budget.allItems,
    totalIncome: state.budget.totals.income,
  }));

  //Hook to calculate total income and save it to store.
  React.useEffect(() => {
    if (allItems.income.length === 0) return;
    let totalIncome = 0;
    for (const income of allItems.income) {
      totalIncome += income.cost;
    }
    if (totalIncome > 0) {
      const payload = { totalIncome };
      dispatch(addTotalIncome(payload));
    }
  }, [allItems.income, dispatch]);

  //Hook to prepare data set for Sankey chart and save it to store.
  React.useEffect(() => {
    if (allItems.income.length === 0 || allItems.expense.length === 0) return;
    let sankeyItemIncomeArr = [];
    let sankeyItemExpenseArr = [];
    let sankeyItem = [];
    for (const income of allItems.income) {
      let sankeyItemIncome = {};
      sankeyItemIncome.from = `${income.description} (${income.cost})`;
      sankeyItemIncome.to = `Total Income (${totalIncome})`;
      sankeyItemIncome.value = income.cost;
      sankeyItemIncomeArr.push(sankeyItemIncome);
    }
    for (const expense of allItems.expense) {
      let sankeyItemExpense = {};
      sankeyItemExpense.to = `${expense.description} (${expense.cost})`;
      sankeyItemExpense.from = `Total Income (${totalIncome})`;
      sankeyItemExpense.value = expense.cost;
      sankeyItemExpenseArr.push(sankeyItemExpense);
    }
    //Concate both the array of objects.
    sankeyItem = [...sankeyItemIncomeArr, ...sankeyItemExpenseArr];
    const payload = { sankeyItem };
    if (sankeyItem.length > 0) {
      dispatch(addDataForSankeyChart(payload));
    }
  }, [allItems.expense, allItems.income, dispatch, totalIncome]);

  //EventHandler for Select box.
  const handleBudgetType = (event) => {
    setFormData({ ...formData, type: event.target.value });
  };

  //EventHandler for Description input field.
  const handleInputDescChange = (event) => {
    setFormData({ ...formData, description: event.target.value });
  };

  //EventHandler for value input field.
  const handleInputValueChange = (event) => {
    setFormData({ ...formData, cost: +event.target.value });
  };

  const addItem = () => {
    let id = 0;
    let newItem = {};
    //Create a copy of form data to be reset.
    const clearFormData = { ...formData };
    //Validation for Input fields.
    if (!description) {
      return alert("Please enter the required value.");
    }
    if (!cost && cost !== 0) {
      return alert("Please enter the required value.");
    }

    //Create an id based on the available items in store.
    if (allItems[type].length > 0) {
      id = allItems[type][allItems[type].length - 1].id + 1;
    }

    //Based on the type of cash flow create respective objects for Income and Expenses.
    if (type === "income") {
      newItem.id = id;
      newItem.description = description;
      newItem.cost = cost;
      newItem.type = type;
    } else {
      newItem.id = id;
      newItem.description = description;
      newItem.cost = cost;
      newItem.type = type;
    }

    //Dispatch an action to add item to the store.
    dispatch(addItemToStore(newItem));
    //Reset inputs.
    setFormData({
      ...clearFormData,
      description: "",
      cost: "",
      type: "income",
    });
  };

  //EventHandler for Add button click.
  const handleAddBtnClick = () => {
    addItem();
  };

  //EventHandler for Keypress
  const handleKeyPressClick = (event) => {
    //To check only if enter key is pressed.
    if (event.keyCode === 13 || event.which === 13) {
      addItem();
    }
  };

  return (
    <div className="bottom">
      <div className="add" onKeyPress={handleKeyPressClick}>
        <div className="add__container">
          <Select
            onChange={handleBudgetType}
            value={type}
            className="add__type"
          />
          <Input
            type="text"
            name="description"
            className="add__description"
            placeholder={t("bottom.description")}
            value={description}
            onChange={handleInputDescChange}
          />
          <Input
            type="number"
            name="value"
            className="add__value"
            placeholder={t("bottom.value")}
            value={cost}
            onChange={handleInputValueChange}
          />
          <Button
            onClick={handleAddBtnClick}
            buttonIcon="ion-ios-checkmark-outline"
            className="add__btn"
          />
        </div>
      </div>
      <div className="container clearfix">
        <BudgetList allItems={allItems} />
      </div>
    </div>
  );
};

export default AddBudget;
