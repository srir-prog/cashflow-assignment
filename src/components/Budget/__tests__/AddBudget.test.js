import React from 'react';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import AddBudget from './../AddBudget';
import BudgetList from './../BudgetList';
import configureStore from 'redux-mock-store';
import Button from './../../Generic/Button';
import Input from './../../Generic/Input';

const mockStore = configureStore({});

let store;

let component;
beforeEach(() => {

    store = mockStore({
        budget: {
            allItems: { income: [], expense: [] },
            totals: { income: 0, expense: 0 }
        }
    });

    component = renderer.create(
        <Provider store={store}>
            <AddBudget />
        </Provider>
    );
});
it('should render correctly', () => {
    //Snapshot testing.
    expect(component.toJSON()).toMatchSnapshot();
});

it('should contains <BudgetList/> component', () => {
    let component = shallow(
        <Provider store={store}>
            <AddBudget />
        </Provider>
    );
    expect(component.containsMatchingElement(<BudgetList />)).toEqual(false);
});

it('should check for add button click', () => {
    const mockAddBtnClick = jest.fn();
    let buttonElement = shallow(<Button onClick={mockAddBtnClick} />);
    buttonElement.find('button').props().onClick();
    expect(mockAddBtnClick).toHaveBeenCalled();
});

it('should check for add description input', () => {
    const mockChangeEvent = jest.fn();
    let inputElement = shallow(<Input onChange={mockChangeEvent} value="Salary" />);
    inputElement.find('input').simulate('change', { target: { value: "Salary" } });
    expect(inputElement.find('input').prop('value')).toEqual("Salary");
});

it('should check for cost input', () => {
    const mockChangeEvent = jest.fn();
    let inputElement = shallow(<Input onChange={mockChangeEvent} value="10" />);
    inputElement.find('input').simulate('change', { target: { value: "10" } });
    expect(inputElement.find('input').prop('value')).toEqual("10");
});