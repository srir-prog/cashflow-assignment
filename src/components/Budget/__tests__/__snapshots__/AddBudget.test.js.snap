// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="bottom"
>
  <div
    className="add"
    onKeyPress={[Function]}
  >
    <div
      className="add__container"
    >
      <select
        className="add__type"
        onChange={[Function]}
        value="income"
      >
        <option
          defaultValue={true}
          value="income"
        >
          bottom.income
        </option>
        <option
          value="expense"
        >
           
          bottom.expense
        </option>
      </select>
      <input
        className="add__description"
        name="description"
        onChange={[Function]}
        placeholder="bottom.description"
        type="text"
        value=""
      />
      <input
        className="add__value"
        name="value"
        onChange={[Function]}
        placeholder="bottom.value"
        type="number"
        value=""
      />
      <button
        className="add__btn"
        onClick={[Function]}
      >
        <i
          className="ion-ios-checkmark-outline"
        />
      </button>
    </div>
  </div>
  <div
    className="container clearfix"
  >
    <div
      className="container clearfix"
    >
      <div
        className="income"
      >
        <h2
          className="icome__title"
        >
          bottom.income
        </h2>
        <div
          className="income__list"
        />
      </div>
      <div
        className="expenses"
      >
        <h2
          className="expenses__title"
        >
          bottom.expenses
        </h2>
        <div
          className="expenses__list"
        />
      </div>
    </div>
  </div>
</div>
`;
